docarray/doc_index/backends/hnswlib_doc_index.py

```diff
@@ -0,0 +1,400 @@
+import hashlib
+import os
+import sqlite3
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Generic,
+    List,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import hnswlib
+import numpy as np
+
+import docarray.typing
+from docarray import BaseDocument, DocumentArray
+from docarray.doc_index.abstract_doc_index import (
+    BaseDocumentIndex,
+    FindResultBatched,
+    _Column,
+)
+from docarray.proto import DocumentProto
+from docarray.utils.filter import filter as da_filter
+from docarray.utils.find import FindResult
+from docarray.utils.misc import is_np_int, torch_imported
+
+if TYPE_CHECKING:
+
+    def composable(fn):  # type: ignore
+        return fn
+
+else:
+    # static type checkers do not like callable objects as decorators
+    from docarray.doc_index.abstract_doc_index import composable
+
+TSchema = TypeVar('TSchema', bound=BaseDocument)
+T = TypeVar('T', bound='HnswDocumentIndex')
+
+HNSWLIB_PY_VEC_TYPES = [list, tuple, np.ndarray]
+if torch_imported:
+    import torch
+
+    HNSWLIB_PY_VEC_TYPES.append(torch.Tensor)
+
+
+class HnswDocumentIndex(BaseDocumentIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._db_config = cast(HnswDocumentIndex.DBConfig, self._db_config)
```

Comment:
can we move this line to `super().__init__` as well? like
`self._db_config = cast(self.DBConfig, self._db_config)`

Reply:
No I don't think so. The issue is that mypy wants `self._db_config` to be of type `HnswDocumentIndex.DBConfig`, and not of type `AbstractDocumentIndex.DBConfig`, and the abstract class (`super()`) does not know about `HnswDocumentIndex.DBConfig`.
But luckily this is only about mypy, the code will run perfectly fine also without this line